### PR TITLE
Refactor SchemeWithRetries in curl package

### PR DESCRIPTION
**Pull request by rjoleary**
SchemeWithRetries is now a bit more generic so it works for both HTTP
and TFTP. You can filter the retry on contents of the error message.

Signed-off-by: Ryan O'Leary <ryanoleary@google.com>

### CODE REVIEW

**Comment by rjoleary** [comment]
Let me know if this is designed well. I'll add some unit tests.

File: pkg/curl/schemes.go [hunk 1]
```diff
@@ -43,6 +44,15 @@ type FileScheme interface {
 	Fetch(u *url.URL) (io.ReaderAt, error)
 }
 
+// FileSchemeRetryFilter contains extra RetryFilter method for a FileScheme
+// wrapped by SchemeWithRetries.
+type FileSchemeRetryFilter interface {
+	// RetryFilter lets a FileScheme filter for errors returned by Fetch
+	// which are worth retrying. If this interface is not implemented, the
+	// default for SchemeWithRetries is to always retry.
+	RetryFilter(u *url.URL, err error) bool
```
Comment (hugelgupf) [inline pkg/curl/schemes.go]: why does this take the url?

**Comment by rjoleary** [comment]
Tests added. I uncovered a bug in the existing tests (or at least very confusing logic). The `return` on this line https://github.com/u-root/u-root/blob/d5c91c6659630db026b628346ff6400d2010e9a7/pkg/curl/schemes_test.go#L87 should be a `continue` otherwise it skips testing `LazyFetch` on the next round. After fixing this, I decided to split Fetch and LazyFetch into two separate tests since they significantly differed in where the errors and counts are recorded.

File: pkg/curl/schemes.go (hunk 1 above)
Comment (rjoleary) [reply pkg/curl/schemes.go]: Look at the implementation `func (s Schemes) RetryFilter(u *url.URL, err error)`. `Schemes` looks at the url to find the appropriate FileScheme and calls `RetryFilter` on it with the same arguments.

Comment (hugelgupf) [reply pkg/curl/schemes.go]: can you annotate in the comment what true and false mean?

RetryFilters return true to mean the request should be retried.

Comment (rjoleary) [reply pkg/curl/schemes.go]: Done.

**Comment by hugelgupf** [comment]
why did this never automerge? weird automation...

**Pull request merged**